### PR TITLE
Discover relations for Oracle provider

### PR DESCRIPTION
**Pull request by Djedouas**
## Description

This PR adds the discovery of foreign key relations for Oracle provider.

I based my work on the way Postgres provider discovers foreign key relations, and adapted the code and the tests to fit Oracle.

See #53769 for the PR about Postgres provider minor simplifications.

## Snapshot

image

---

cc @Jean-Roc 

Funded by Métropole de Lille

### CODE REVIEW

File: src/providers/oracle/qgsoracleprovider.cpp [hunk 1]
```diff
@@ -2505,6 +2506,118 @@ bool QgsOracleProvider::setSubsetString( const QString &theSQL, bool updateFeatu
...
+        }
+        else
+        {
+          QgsLogger::warning( "Invalid relation for " + name + ", validation error: " + relation.validationError() );
+        }
+      }
+    }
+    else
+    {
+      // multi reference field => add the field pair to all the referenced layers found
```
Comment (troopa81) [inline src/providers/oracle/qgsoracleprovider.cpp:2604]: Same instruction than branch above, could we not factorize it before if ?

File: src/providers/oracle/qgsoracleprovider.cpp [hunk 2]
```diff
@@ -2505,6 +2506,118 @@ bool QgsOracleProvider::setSubsetString( const QString &theSQL, bool updateFeatu
...
+        QgsRelation relation;
+        relation.setName( name );
+        relation.setReferencingLayer( target->id() );
+        relation.setReferencedLayer( foundLayer->id() );
+        relation.addFieldPair( fkColumn, refColumn );
+        relation.generateId();
+        if ( relation.isValid() )
+        {
+          result.append( relation );
+          ++nbFound;
```
Comment (troopa81) [inline src/providers/oracle/qgsoracleprovider.cpp:2593]: nbFound is useless no? refTableFound is enough

File: src/providers/oracle/qgsoracleprovider.cpp [hunk 3]
```diff
@@ -2505,6 +2506,118 @@ bool QgsOracleProvider::setSubsetString( const QString &theSQL, bool updateFeatu
...
+      // multi reference field => add the field pair to all the referenced layers found
+      const QList<QgsVectorLayer *> foundLayers = searchLayers( layers, mUri.connectionInfo( false ), refOwner, refTable );
+      for ( int i = 0; i < nbFound; ++i )
+      {
+        for ( const QgsVectorLayer *foundLayer : foundLayers )
+        {
+          if ( result[result.size() - 1 - i].referencedLayerId() == foundLayer->id() )
+          {
+            result[result.size() - 1 - i].addFieldPair( fkColumn, refColumn );
+          }
```
Comment (troopa81) [inline src/providers/oracle/qgsoracleprovider.cpp:2613]: else what? If we have a multi ref field to a table existing in several layers, it wouldn't work correctly no?

**Comment by Gustry** [comment]
@Djedouas I just understand why, but you shouldn't add the backtick in your funded by description... It won't by parsed. Just write it as text ;-)

**Review by troopa81 — APPROVED** [review]

File: src/providers/oracle/qgsoracleprovider.cpp [hunk 4]
```diff
@@ -2505,6 +2506,118 @@ bool QgsOracleProvider::setSubsetString( const QString &theSQL, bool updateFeatu
...
+        else
+        {
+          QgsLogger::warning( "Invalid relation for " + name + ", validation error: " + relation.validationError() );
+        }
+      }
+    }
+    else
+    {
+      // multi reference field => add the field pair to all the referenced layers found
+      int resultSize = result.size();
```
Comment (troopa81) [inline src/providers/oracle/qgsoracleprovider.cpp:2605]:
```suggestion
      const int resultSize = result.size();
```

**Pull request merged**